src/shogun2-core/shogun2-service/src/main/java/de/terrestris/shogun2/service/GeoServerInterceptorService.java

```diff
@@ -0,0 +1,611 @@
+			GeoServerInterceptorService.class);
+
+	/**
+	 * The autowired properties file containing the (application driven)
+	 * GeoServer namespace - GeoServer BaseURI mapping, e.g.:
+	 *
+	 *   topp=http://localhost:8080/geoserver/topp/ows
+	 *
+	 */
+	private Properties geoServerNameSpaces;
+
+	/**
+	 * An array of whitelisted Headers to forward within the Interceptor.
+	 */
+	private static final String[] FORWARD_HEADER_KEYS = new String[] {
+		"Content-Type",
+		"Content-Disposition",
+		"geowebcache-cache-result",
+		"geowebcache-crs",
+		"geowebcache-gridset",
+		"geowebcache-tile-bounds",
+		"geowebcache-tile-index"
+	};
+
+	/**
+	 *
+	 */
+	@Autowired
+	OgcMessageDistributor ogcMessageDistributor;
+
+	/**
+	 *
+	 */
+	@Autowired
+	InterceptorRuleService<InterceptorRule, ?> interceptorRuleService;
+
+	/**
+	 *
+	 * @param request
+	 * @return
+	 * @throws InterceptorException
+	 * @throws URISyntaxException
+	 * @throws UnsupportedEncodingException
+	 * @throws HttpException
+	 */
+	public Response interceptGeoServerRequest(HttpServletRequest request)
+			throws InterceptorException, URISyntaxException,
+			UnsupportedEncodingException, HttpException {
+
+		// wrap the request, we want to manipulate it
+		MutableHttpServletRequest mutableRequest =
+				new MutableHttpServletRequest(request);
+
+		// get the OGC request information (service, request, endPoint)
+		OgcMessage message = getOgcRequest(mutableRequest);
+
+		// get the GeoServer base URI by the provided request
+		URI geoServerBaseUri = getGeoServerBaseURI(message);
+
+		// set the GeoServer base URI to the (wrapped) request
+		mutableRequest.setRequestURI(geoServerBaseUri);
+
+		// intercept the request (if needed)
+		mutableRequest = ogcMessageDistributor
+				.distributeToRequestInterceptor(mutableRequest, message);
+
+		// send the request
+		// TODO: Move to global proxy class
+		Response response = sendRequest(mutableRequest);
+
+		// intercept the response (if needed)
+		Response interceptedResponse = ogcMessageDistributor
+				.distributeToResponeInterceptor(response, message);
+
+		// finally filter the white-listed response headers
+		// TODO: Move to global proxy class
+		interceptedResponse.setHeaders(
+				getHeadersToForward(interceptedResponse.getHeaders()));
+
+		return interceptedResponse;
+	}
+
+	/**
+	 *
+	 * @param mutableRequest
+	 * @return
+	 * @throws InterceptorException
+	 */
+	private OgcMessage getOgcRequest(MutableHttpServletRequest mutableRequest)
+			throws InterceptorException {
+
+		OgcMessage ogcRequest = new OgcMessage();
+
+		String requestService = getRequestParameterValue(
+				mutableRequest, OgcNaming.PARAMETER_SERVICE);
+		String requestOperation = getRequestParameterValue(
+				mutableRequest, OgcNaming.PARAMETER_OPERATION);
+		String requestLayer = getRequestParameterValue(
+				mutableRequest, OgcNaming.PARAMETER_ENDPOINT);
+		InterceptorRule mostSpecificRequestRule = getMostSpecificRule(requestService,
```

Comment:
I would set the `service`, `operation` and `endpoint` for the `OgcMessage ogcRequest` and pass this then on to the method `getMostSpecificRule`, instead of sending the strings.